File: packages/go/realtimemail/gateway.go
package realtimemail

import (
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"strings"
	"time"
)

type HostActionBroker struct {
	TrustPolicy *TrustPolicy
	Verifier    SignatureVerifier
}

func (broker HostActionBroker) Authorize(action RealtimeMailAction, message RealtimeMailMessage, manifest RealtimeMailManifest, userGesture bool, now time.Time) (RealtimeMailAction, error) {
	if issues := (ActionValidator{}).Validate(action); len(issues) > 0 {
		return action, errors.New("invalid_action")
	}
	if action.MessageID != message.ID {
		return action, errors.New("message_mismatch")
	}
	if action.Domain != message.Domain || action.Domain != manifest.Domain {
		return action, errors.New("domain_mismatch")
	}
	if broker.TrustPolicy == nil || !broker.TrustPolicy.IsTrusted(action.Domain) {
		return action, errors.New("domain_not_trusted")
	}
	if !userGesture || !action.RequiresUserGesture {
		return action, errors.New("user_gesture_required")
	}
	if !message.ExpiresAt.IsZero() && !message.ExpiresAt.After(now) {
		return action, errors.New("message_expired")
	}
	if !broker.verifyWithManifestKey(message, manifest) {
		return action, errors.New("signature_required")
	}
	if action.Type == OpenURL && !hasCapability(message.Capabilities, OpenURLUserGesture) {
		return action, errors.New("capability_required")
	}
	if isPaymentRequest(action.Payload) && !hasCapability(message.Capabilities, PaymentRequestUserGesture) {
		return action, errors.New("capability_required")
	}
	return action, nil
}

func (broker HostActionBroker) verifyWithManifestKey(message RealtimeMailMessage, manifest RealtimeMailManifest) bool {
	for _, key := range manifest.PublicKeys {
		if strings.HasPrefix(key, "ed25519:") && broker.Verifier.VerifyEd25519(message, key) {
			return true
		}
	}
	return false
}

type RealtimeMessageBuilder struct {
	Manifest RealtimeMailManifest
	Clock    func() time.Time
}

func (builder RealtimeMessageBuilder) Build(input RealtimeMessageInput) (RealtimeMailMessage, error) {
	channel, ok := builder.channel(input.ChannelID)
	if !ok {
		return RealtimeMailMessage{}, errors.New("unknown_channel")
	}
	clock := builder.Clock
	if clock == nil {
		clock = time.Now
	}
	capabilities := input.Capabilities
	if capabilities == nil {
		capabilities = channel.Capabilities
	}
	message := RealtimeMailMessage{
		ID:           input.ID,
		Source:       Realtime,
		Domain:       builder.Manifest.Domain,
		ChannelID:    channel.ID,
		From:         input.From,
		Subject:      input.Subject,
		HTML:         input.HTML,
		CSS:          input.CSS,
		Script:       input.Script,
		Capabilities: capabilities,
		ReceivedAt:   clock(),
		ExpiresAt:    input.ExpiresAt,
	}
	if message.ID == "" {
		message.ID = "generated-message"
	}
	validationMessage := message
	validationMessage.Signature = "unsigned-builder-placeholder"
	if issues := (MessageValidator{}).Validate(validationMessage); len(issues) > 0 {
		return message, ValidationError{Issues: issues}
	}
	message.Signature = ""
	return message, nil
}

func (builder RealtimeMessageBuilder) channel(id string) (RealtimeMailChannel, bool) {
	for _, channel := range builder.Manifest.Channels {
		if channel.ID == id {
			return channel, true
		}
	}
	return RealtimeMailChannel{}, false
}

type RealtimeMessageInput struct {
	ID           string
	ChannelID    string
	From         string
	Subject      string
	HTML         string
	CSS          string
	Script       string
	Capabilities []TrustCapability
	ExpiresAt    time.Time
}

type MessageSigner struct {
	Verifier SignatureVerifier
}

func (signer MessageSigner) SignEd25519(message RealtimeMailMessage, privateKey ed25519.PrivateKey) (RealtimeMailMessage, error) {
	message.Signature = ""
	canonical, err := signer.Verifier.CanonicalMessage(message)
	if err != nil {
		return message, err
	}
	signature := ed25519.Sign(privateKey, []byte(canonical))
	message.Signature = "rmail1.eyJhbGciOiJFZDI1NTE5IiwidHlwIjoicm1haWwxIn0." + base64.RawURLEncoding.EncodeToString(signature)
	return message, nil
}

type RouteAuthorizer struct {
	Manifest RealtimeMailManifest
}

func (authorizer RouteAuthorizer) Authorize(route string, channelID string, userID string) (RealtimeMailChannel, error) {
	for _, channel := range authorizer.Manifest.Channels {
		if channelID != "" && channel.ID != channelID {
			continue
		}
		if routeMatches(channel.Route, route, userID) {
			return channel, nil
		}
	}
	return RealtimeMailChannel{}, errors.New("route_not_allowed")
}

type ActionReceiver struct {
	Domain string
}

func (receiver ActionReceiver) Receive(action RealtimeMailAction) (RealtimeMailAction, error) {
	if issues := (ActionValidator{}).Validate(action); len(issues) > 0 {
		return action, errors.New("invalid_action")
	}
	if action.Domain != receiver.Domain {
		return action, errors.New("domain_not_allowed")
	}
	return action, nil
}

func routeMatches(pattern string, route string, userID string) bool {
	patternParts := splitRoute(pattern)
	routeParts := splitRoute(route)
	if len(patternParts) != len(routeParts) {
		return false
	}
	for index, patternPart := range patternParts {
		routePart := routeParts[index]
		if patternPart == ":userId" {
			if userID == "" || routePart != userID {
				return false
			}
		} else if strings.HasPrefix(patternPart, ":") {
			return false
		} else if patternPart != routePart {
			return false
		}
	}
	return true
}

func splitRoute(route string) []string {
	parts := strings.Split(route, "/")
	filtered := make([]string, 0, len(parts))
	for _, part := range parts {
		if part != "" {
			filtered = append(filtered, part)
		}
	}
	return filtered
}

func isPaymentRequest(payload any) bool {
	_, ok := paymentPayloadMap(payload)
	return ok
}

type PaymentRequestSecurityContext struct {
	Action                       RealtimeMailAction
	Message                      RealtimeMailMessage
	Manifest                     RealtimeMailManifest
	SourceMatchesSelectedSandbox bool
	ExpectedInvoiceID            string
	ExpectedAmount               string
	ExpectedCurrency             string
	ProcessedInvoiceIDs          []string
	Now                          time.Time
}

type PaymentRequestSecurityDecision struct {
	OK      bool
	Reason  string
	Payload map[string]any
}

type PaymentRequestSecurityPolicy struct{}

func (PaymentRequestSecurityPolicy) Authorize(context PaymentRequestSecurityContext) PaymentRequestSecurityDecision {
	if !context.SourceMatchesSelectedSandbox {
		return rejectedPayment("untrusted_frame_source")
	}
	payload, ok := paymentPayloadMap(context.Action.Payload)
	if context.Action.Type != PublishGatewayEvent || !ok {
		return rejectedPayment("payment_payload_required")
	}
	if len((PaymentRequestPayloadValidator{}).Validate(payload)) > 0 {
		return rejectedPayment("invalid_payment_payload")
	}
	if context.Action.MessageID != context.Message.ID {
		return rejectedPayment("message_mismatch")
	}
	if context.Action.Domain != context.Message.Domain || context.Action.Domain != context.Manifest.Domain {
		return rejectedPayment("domain_mismatch")
	}
	merchant := payload["merchant"].(map[string]any)
	if asString(merchant["domain"]) != context.Message.Domain {
		return rejectedPayment("merchant_domain_mismatch")
	}
	if !hasCapability(context.Message.Capabilities, PaymentRequestUserGesture) {
		return rejectedPayment("capability_required")
	}
	if context.ExpectedInvoiceID != "" && asString(payload["invoiceId"]) != context.ExpectedInvoiceID {
		return rejectedPayment("invoice_mismatch")
	}
	amount := payload["amount"].(map[string]any)
	if context.ExpectedAmount != "" && asString(amount["value"]) != context.ExpectedAmount {
		return rejectedPayment("amount_mismatch")
	}
	if context.ExpectedCurrency != "" && asString(amount["currency"]) != context.ExpectedCurrency {
		return rejectedPayment("currency_mismatch")
	}
	expiresAt, err := time.Parse(time.RFC3339, asString(payload["expiresAt"]))
	if err != nil || !expiresAt.After(context.Now) {
		return rejectedPayment("payment_expired")
	}
	if containsString(context.ProcessedInvoiceIDs, asString(payload["invoiceId"])) {
		return rejectedPayment("duplicate_invoice")
	}
	return PaymentRequestSecurityDecision{OK: true, Reason: "ok", Payload: payload}
}

func rejectedPayment(reason string) PaymentRequestSecurityDecision {
	return PaymentRequestSecurityDecision{OK: false, Reason: reason}
}

File: packages/python/src/realtime_mail/gateway.py
from __future__ import annotations

from dataclasses import replace
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from .models import MailSource, RealtimeMailAction, RealtimeMailManifest, RealtimeMailMessage, TrustCapability, action_from_dict
from .signature import SignatureVerifier
from .trust import TrustPolicy
from .validation import ActionValidator, ValidationError, ValidationIssue, MessageValidator


class HostActionBroker:
    def __init__(self, trust_policy: TrustPolicy, verifier: SignatureVerifier | None = None) -> None:
        self.trust_policy = trust_policy
        self.verifier = verifier or SignatureVerifier()

    def authorize(
        self,
        action_value: Any,
        message: RealtimeMailMessage,
        manifest: RealtimeMailManifest,
        user_gesture: bool,
        now: datetime | None = None,
    ) -> tuple[bool, str, RealtimeMailAction | None]:
        issues = ActionValidator.validate(action_value)
        if issues:
            return False, "invalid_action", None
        action = action_from_dict(action_value)
        if action.message_id != message.id:
            return False, "message_mismatch", None
        if action.domain != message.domain or action.domain != manifest.domain:
            return False, "domain_mismatch", None
        if not self.trust_policy.is_trusted(action.domain):
            return False, "domain_not_trusted", None
        if not user_gesture or not action.requires_user_gesture:
            return False, "user_gesture_required", None
        if message.expires_at and message.expires_at <= (now or datetime.now(timezone.utc)):
            return False, "message_expired", None
        if not any(key.startswith("ed25519:") and self.verifier.verify_ed25519(message, key) for key in manifest.public_keys):
            return False, "signature_required", None
        if action.type.value == "open_url" and TrustCapability.OPEN_URL_USER_GESTURE not in message.capabilities:
            return False, "capability_required", None
        if _is_payment_request(action.payload) and TrustCapability.PAYMENT_REQUEST_USER_GESTURE not in message.capabilities:
            return False, "capability_required", None
        return True, "ok", action


class RealtimeMessageBuilder:
    def __init__(self, manifest: RealtimeMailManifest, clock=datetime.now) -> None:
        self.manifest = manifest
        self.clock = clock

    def build(
        self,
        *,
        channel_id: str,
        from_address: str,
        subject: str,
        html: str,
        css: str | None = None,
        script: str | None = None,
        capabilities: list[TrustCapability] | None = None,
        expires_at: datetime | None = None,
        id: str | None = None,
    ) -> RealtimeMailMessage:
        channel = next((candidate for candidate in self.manifest.channels if candidate.id == channel_id), None)
        if channel is None:
            raise ValueError(f"Unknown channel: {channel_id}")
        message = RealtimeMailMessage(
            id=id or str(uuid4()),
            source=MailSource.REALTIME,
            domain=self.manifest.domain,
            channel_id=channel.id,
            from_address=from_address,
            subject=subject,
            html=html,
            css=css,
            script=script,
            capabilities=capabilities or channel.capabilities,
            received_at=self.clock(timezone.utc),
            expires_at=expires_at,
        )
        issues = MessageValidator.validate({**_message_to_wire(message), "signature": "unsigned-builder-placeholder"})
        if issues:
            raise ValidationError(issues)
        return message


class MessageSigner:
    def __init__(self, verifier: SignatureVerifier | None = None) -> None:
        self.verifier = verifier or SignatureVerifier()

    def sign_ed25519(self, message: RealtimeMailMessage, private_key: Any) -> RealtimeMailMessage:
        import base64

        signature = private_key.sign(self.verifier.canonical_message(message).encode("utf-8"))
        encoded = base64.urlsafe_b64encode(signature).decode("ascii").rstrip("=")
        return replace(message, signature=f"rmail1.eyJhbGciOiJFZDI1NTE5IiwidHlwIjoicm1haWwxIn0.{encoded}")


class RouteAuthorizer:
    def __init__(self, manifest: RealtimeMailManifest) -> None:
        self.manifest = manifest

    def authorize(self, route: str, channel_id: str | None = None, user_id: str | None = None) -> tuple[bool, str]:
        for channel in self.manifest.channels:
            if channel_id and channel.id != channel_id:
                continue
            if _route_matches(channel.route, route, user_id):
                return True, "ok"
        return False, "route_not_allowed"


class ActionReceiver:
    def __init__(self, domain: str) -> None:
        self.domain = domain

    def receive(self, value: Any) -> tuple[bool, str, RealtimeMailAction | None]:
        issues = ActionValidator.validate(value)
        if issues:
            return False, "invalid_action", None
        action = action_from_dict(value)
        if action.domain != self.domain:
            return False, "domain_not_allowed", None
        return True, "ok", action


def _route_matches(pattern: str, route: str, user_id: str | None) -> bool:
    pattern_parts = [part for part in pattern.split("/") if part]
    route_parts = [part for part in route.split("/") if part]
    if len(pattern_parts) != len(route_parts):
        return False
    for pattern_part, route_part in zip(pattern_parts, route_parts):
        if pattern_part == ":userId":
            if not user_id or route_part != user_id:
                return False
        elif pattern_part.startswith(":"):
            continue
        elif pattern_part != route_part:
            return False
    return True


def _is_payment_request(payload: object | None) -> bool:
    return isinstance(payload, dict) and payload.get("kind") == "host-mediated-payment-request"


def _message_to_wire(message: RealtimeMailMessage) -> dict[str, Any]:
    return {
        "id": message.id,
        "source": message.source.value,
        "domain": message.domain,
        "channelId": message.channel_id,
        "from": message.from_address,
        "subject": message.subject,
        "html": message.html,
        "css": message.css,
        "script": message.script,
        "capabilities": [capability.value for capability in message.capabilities],
        "receivedAt": message.received_at.isoformat().replace("+00:00", "Z"),
        "expiresAt": message.expires_at.isoformat().replace("+00:00", "Z") if message.expires_at else None,
    }
